worker/wallet.js

```diff
@@ -255,14 +255,34 @@ async function checkWithdrawal ({ data: { hash }, boss, models, lnd }) {
   }
 }
 
-export async function autoDropBolt11s ({ models }) {
-  await serialize(models, models.$executeRaw`
-    UPDATE "Withdrawl"
-    SET hash = NULL, bolt11 = NULL
-    WHERE "userId" IN (SELECT id FROM users WHERE "autoDropBolt11s")
-    AND now() > created_at + interval '${INVOICE_RETENTION_DAYS} days'
-    AND hash IS NOT NULL;`
-  )
+export async function autoDropBolt11s ({ models, lnd }) {
+  const invoices = await serialize(models, models.$executeRaw`
+    WITH updated_rows AS (
+      UPDATE "Withdrawl"
+      SET hash = NULL, bolt11 = NULL
+      WHERE "userId" IN (SELECT id FROM users WHERE "autoDropBolt11s")
+      AND now() > created_at + interval '${INVOICE_RETENTION_DAYS} days'
+      AND hash IS NOT NULL
+      RETURNING id, hash, bolt11
+    )
+    SELECT * FROM updated_rows;
+  `)
+
+  const failedDeletesUpdatePromises = []
+  for (const invoice of invoices) {
+    try {
+      await deletePayment({ id: invoice.hash, lnd })
+    } catch (error) {
+      console.error(`Error removing invoice with hash ${invoice.hash}:`, error)
+      failedDeletesUpdatePromises.push(models.withdrawl.update({
+        where: { id: invoice.id },
+        data: { hash: invoice.hash, bolt11: invoice.bolt11 }
+      }))
+    }
+  }
```

Comment:
`UPDATE ... RETURNING` returns the updated values not the old ones. So this wouldn't work since `invoice.hash` and `invoice.bolt11` is null now:

> In an UPDATE, the data available to RETURNING is the new content of the modified row.

-- https://www.postgresql.org/docs/13/dml-returning.html

I can see how my previous comment was misleading though since I mentioned that "the existing code was fine". I wanted to say that the end result for the withdrawal table should be the same as before.

Reply:
The `WITH` statement should be making a CTE that returns the values before they were updated

[https://www.postgresql.org/docs/current/queries-with.html](url) But looking at the example at the bottom I might have written it wrong lol

Reply:
> The `WITH` statement should be making a CTE that returns the values before they were updated

You're right that it creates a CTE and a snapshot of the database and individual CTE's don't see the result of each other but afaict, this doesn't mean that `UPDATE ... RETURNING` inside a CTE does return the original rows. For example, I just tested the behavior on my local machine (using the postgres:13.2 image) with the following queries:

```
stackernews=# SELECT id, hash FROM "Withdrawl" ORDER BY created_at DESC LIMIT 3;
  id   |                               hash
-------+------------------------------------------------------------------
 18326 | 5ed899dc1f0c43c34927e249a4c469a137f4974ce034fa87c9cb38eca7dc6aa6
 18325 | 469b477f103a798d41e7561ad5d0e44f4e1657a78738a1b9546695abe8be7e46
 18324 | 0bcefb7d980e66bad8fd57449a474bf3c813ca28e46dd75c646041dd1ea70110
(3 rows)

stackernews=# WITH updated_rows AS ( UPDATE "Withdrawl" SET hash = NULL WHERE id IN (18326, 18325, 18324) RETURNING id, hash ) SELECT * FROM updated_rows;
  id   | hash
-------+------
 18324 |
 18325 |
 18326 |
(3 rows)
```

> But looking at the example at the bottom I might have written it wrong lol

Which example do you mean?

Reply:
the last 2 on that page. I ended up using the wrong query so I updated it and pushed

Reply:
Ah, I see. Interesting, didn't know about that!